OpenSim/Common/SmoothSegmentedFunction.cpp

```diff
@@ -582,59 +583,98 @@ If x is in the linear region
+    const double x1 = smoothData->_x1;
 
     if (x < x0) {
-        const double y0 = _smoothData->_y0;
-        const double dydx0 = _smoothData->_dydx0;
-        switch (order) {
-            case 0: return y0 + dydx0*(x-x0);
-            case 1: return dydx0;
-            default: return 0.;
-        }
+        const double y0    = smoothData->_y0;
+        const double dydx0 = smoothData->_dydx0;
+        return {y0 + dydx0 * (x - x0), dydx0};
     }
 
+    if (x > x1) {
+        const double y1    = smoothData->_y1;
+        const double dydx1 = smoothData->_dydx1;
+        return {y1 + dydx1 * (x - x1), dydx1};
+    }
+
+    DerivativeValues y{};
     if (x <= x1) {
-        const SimTK::Array_<SimTK::Vec6>& ctrlPtsX = _smoothData->_ctrlPtsX;
-        const int idx  = SegmentedQuinticBezierToolkit::calcIndex(x,ctrlPtsX);
+        const SimTK::Array_<SimTK::Vec6>& ctrlPtsX = smoothData->_ctrlPtsX;
+        const int idx = SegmentedQuinticBezierToolkit::calcIndex(x, ctrlPtsX);
 
         const SimTK::Array_<SimTK::Spline>& arraySplineUX =
-            _smoothData->_arraySplineUX;
+            smoothData->_arraySplineUX;
         const double u = SegmentedQuinticBezierToolkit::calcU(
-                x,
-                ctrlPtsX[idx],
-                arraySplineUX[idx],
-                UTOL,
-                MAXITER);
-
-        const SimTK::Array_<SimTK::Vec6>& ctrlPtsY = _smoothData->_ctrlPtsY;
-        return SegmentedQuinticBezierToolkit::calcQuinticBezierCurveDerivDYDX(
-                u,
-                ctrlPtsX[idx],
-                ctrlPtsY[idx],
-                order);
-    }
-
-    if (x > x1) {
-        const double y1 = _smoothData->_y1;
-        const double dydx1 = _smoothData->_dydx1;
-        switch (order) {
-            case 0: return y1 + dydx1*(x-x1);
-            case 1: return dydx1;
-            default: return 0.;
+            x,
+            ctrlPtsX[idx],
+            arraySplineUX[idx],
+            UTOL,
+            MAXITER);
+
+        const SimTK::Array_<SimTK::Vec6>& ctrlPtsY = smoothData->_ctrlPtsY;
+        for (size_t i = 0; i < y.size(); ++i) {
+            if (selectedOrders[i]) {
+                y.at(i) = SegmentedQuinticBezierToolkit::
+                    calcQuinticBezierCurveDerivDYDX(
+                        u,
+                        ctrlPtsX[idx],
+                        ctrlPtsY[idx],
+                        i);
+            }
         }
+        return y;
     }
 
     // In case of NaN return NaN.
-    return SimTK::NaN;
+    y.fill(SimTK::NaN);
+    return y;
+}
+
+} // namespace
+
+double SmoothSegmentedFunction::calcDerivative(double x, int order) const
+{
+    SelectedDerivativeOrders orders{};
+    orders.at(order) = true;
+    return calcSelectedDerivatives(x, orders, _smoothData).at(order);
+}
+
+SmoothSegmentedFunction::ValueAndDerivative SmoothSegmentedFunction::
+    calcValueAndFirstDerivative(double x) const
+{
+    const SelectedDerivativeOrders orders{true, true};
```

Comment:
UB? (`true`, `true`, UB, UB, etc.)

Reply:
According to c++, when using list initializer the missing values are value initialized. When using `y = {}` it is UB.

Example from cppreference on `std::array`

```
std::array<int, 2> a7{1}; // List init, unspecified element is value
                                               // initialized, a7[0] = 1, a7[1] = 0
```

So it is ok, no?

Reply:
My bad, what you are saying is ok:

- https://en.cppreference.com/w/cpp/language/aggregate_initialization

> For a non-union aggregate, each element that is not an explicitly initialized element is initialized as follows:
>
> - If the element has a default member initializer, the element is initialized from that initializer. | (since C++11)
> - Otherwise, if the element is not a reference, the element is copy-initialized from an empty initializer list.
> - Otherwise, the program is ill-formed.

The same document states that `y = {}`, would mean "no explicitly initialized elements" (i.e. they are all value-initialized as above):

> Determine the <i>explicitly initialized elements</i> of the aggregate as follows:
>
> - If the initializer list is a designated initializer list  (the aggregate can only be of class type), the identifier in each  designator shall name a direct non-static data member of the class, and  the explicitly initialized elements of the aggregate are the elements  that are, or contain, those members. 
> - Otherwise, if the initializer list is non-empty, the explicitly initialized elements of the aggregate are the first <i>n</i> elements of the aggregate, where <i>n</i> is the number of elements in the initializer list.
> - Otherwise, the initializer list must be empty (<span class="t-c"><span class="mw-geshi cpp source-cpp"><span class="br0">{</span><span class="br0">}</span></span></span>), and there are no explicitly initialized elements.


The only thing to be aware of when returning different objects is, rather than one, is:

- https://en.cppreference.com/w/cpp/language/copy_elision

Which isn't going to be important for `double`s